### PR TITLE
Replace `incomplete_beta` with `at.betainc` and speedup/clean logcdf tests

### LINKED ISSUE #4420: dist_math.incomplete_beta is very slow
I notice this when running test_distributions.py. Any distribution `logcdf` method that uses `incomplete_beta` takes ages to run. 

Right now it is used in the `logcdf` methods of  `Beta`, `StudentT`, `Binomial`, `NegativeBinomial`, and `ZeroInflated` versions of the last two distributions.

Is there a reason why this cannot be implemented as a Theano C Op? Scipy has C implementation here, which seems to be doing exactly the same as our algorithm in `dist_math.py`:
https://github.com/scipy/scipy/blob/master/scipy/special/cephes/incbet.c

In addition, this would probably make it trivial to implement a vectorized version for tensors, allowing the logcdf methods of the distributions above to evaluate more than one value at a time.

### LINKED ISSUE #4467: Remove hacks to avoid previous aesara gammainc(c) limitations
**This is a reminder issue**

The gamma functions in aesara were changed to work with invalid values (returning nan) instead of doing a c-level assertion (https://github.com/pymc-devs/aesara/pull/290). The following lines can be changed / removed once pymc3 is paired with the next version of aesara:

Distributions:

https://github.com/pymc-devs/pymc3/blob/89916ad724ae88ebb4c1af3d65919b87b318f180/pymc3/distributions/continuous.py#L2631-L2634

https://github.com/pymc-devs/pymc3/blob/89916ad724ae88ebb4c1af3d65919b87b318f180/pymc3/distributions/continuous.py#L2805-L2808

Unittests:

https://github.com/pymc-devs/pymc3/blob/e40304b98503907f620d67d4a9ff6e42c63171d1/pymc3/tests/test_distributions.py#L1231-L1240

https://github.com/pymc-devs/pymc3/blob/e40304b98503907f620d67d4a9ff6e42c63171d1/pymc3/tests/test_distributions.py#L1253-L1262

### PR DESCRIPTION
**Pull request by ricardoV94**
# Implement Betainc

This PR deprecates the `incomplete_beta` method in `dist_math.py` in favor of the new `aesara.tensor.betainc` which wraps the `scipy.special.betainc` method. Supersedes #4736; Closes #4420.

This change provides a large increase in speed (and graph compilation) for distributions that made use of the `incomplete_beta` and/or it's gradients. In addition, the logcdf of Beta, StudentT, Binomial, NegativeBinomial (and zero inflated variants) can now evaluate multiple values. 

The downside is the constant warning that the op has no `c-implementation` whenever the logp/logcdf makes use of it. This limitation is tracked in https://github.com/aesara-devs/aesara/issues/267 and https://github.com/aesara-devs/aesara/issues/514 

# Speedup logcdf tests

The logcdf tests were refactored to avoid recreating the logcdf graph for every parameter / value evaluation. This supersedes #4734. The old `incomplete_beta` was found to be defective when attempting #4734, and so I decided to merge the two PRs into this one instead.

This allows to remove most of the custom `n_samples` limitations, and makes the test suite run considerably faster.

# Other changes
1. The work-around for the `HyperGeometric` tests caused by https://github.com/aesara-devs/aesara/issues/461 was removed.
1. The work-around in the `Gamma` and `InverseGamma` logcdf methods was removed (closes #4467)

### CODE REVIEW

**Comment by ricardoV94** [comment]
Comparing these two workflows for the `test_distributions.py`: 
1. [Other PR](https://github.com/pymc-devs/pymc3/runs/3042805582?check_suite_focus=true) - 48m
2. [This PR](https://github.com/pymc-devs/pymc3/runs/3045311114?check_suite_focus=true) - 15m

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -3,12 +3,13 @@
 ## PyMC3 vNext (4.0.0)
 ### Breaking Changes
 - ⚠ Theano-PyMC has been replaced with Aesara, so all external references to `theano`, `tt`, and `pymc3.theanof` need to be replaced with `aesara`, `at`, and `pymc3.aesaraf` (see [4471](https://github.com/pymc-devs/pymc3/pull/4471)).
+- ⚠ PyMC3 now requires Scipy version `>= 1.4.1` (see [4736](https://github.com/pymc-devs/pymc3/pull/4736)).
```
Comment (ricardoV94) [inline RELEASE-NOTES.md:6]:
```suggestion
- ⚠ PyMC3 now requires Scipy version `>= 1.4.1` (see [4736](https://github.com/pymc-devs/pymc3/pull/4857)).
```

File: pymc3/distributions/continuous.py [hunk 2]
```diff
@@ -1246,23 +1245,19 @@ def logcdf(value, alpha, beta):
 
         Parameters
         ----------
-        value: numeric
-            Value(s) for which log CDF is calculated.
+        value: numeric or np.ndarray or aesara.tensor
+            Value(s) for which log CDF is calculated. If the log CDF for multiple
+            values are desired the values must be provided in a numpy array or Aesara tensor.
```
Comment (brandonwillard) [inline pymc3/distributions/continuous.py:1250]: Might as well keep this parameter description simple and put all the type-based information in type annotations.

File: pymc3/tests/test_distributions.py [hunk 3]
```diff
@@ -751,25 +738,30 @@ def check_logcdf(
...
                 value = params.pop("value")
-                with Model() as m:
-                    dist = pymc3_dist("y", **params)
+                # Update shared parameter variables in pymc3_logcdf function
+                for param_name, param_value in params.items():
+                    param_vars[param_name].set_value(param_value)
+                pymc3_eval = pymc3_logcdf({"value": value})
+
                 params["value"] = value  # for displaying in err_msg
-                with aesara.config.change_flags(on_opt_error="raise", mode=Mode("py")):
```
Comment (brandonwillard) [inline pymc3/tests/test_distributions.py:766]: Are we setting `on_opt_error="raise"` elsewhere?  This is a good option to have set globally.

Also, are we sure we want to compile all these tests via C?  The Python-based graphs can sometimes be about as quick as the C-compiled graphs + their compilation time.

**Review by brandonwillard — COMMENTED** [review]
Minor comments; otherwise, looks fine&mdash;although it is missing coverage for the deprecation warning.

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:1250]: Yeah, these docstrings could benefit from an overhaul (or be removed altogether). I simply changed them to be equal to those of the remaining distributions.

File: pymc3/tests/test_distributions.py (hunk 3 above)
Comment (ricardoV94) [reply pymc3/tests/test_distributions.py:766]: > Are we setting on_opt_error="raise" elsewhere? This is a good option to have set globally.

I don't think we are. You mean set it up for all the tests?

> Also, are we sure we want to compile all these tests via C? The Python-based graphs can sometimes be about as quick as the C-compiled graphs + their compilation time.

I would say it's better to test them with the default back-end, which for the time being is `C`

**Comment by ricardoV94** [comment]
> Minor comments; otherwise, looks fine—although it is missing coverage for the deprecation warning.

Added a test to cover the deprecation warning

File: pymc3/distributions/continuous.py (hunk 2 above)
Comment (ricardoV94) [reply pymc3/distributions/continuous.py:1250]: Opened an issue here: https://github.com/pymc-devs/pymc3/issues/4859

**Review by michaelosthege — CHANGES_REQUESTED** [review]
I agree we should have coverage on the deprecation warnings.

And the minimum Aesara version also needs to be changed in `/conda-envs/*.yml`.

**Comment by ricardoV94** [comment]
> I agree we should have coverage on the deprecation warnings.

I already added that in my last force-push.

> And the minimum Aesara version also needs to be changed in `/conda-envs/*.yml`.

It doesn't use the `requirements.txt`?

**Comment by michaelosthege** [comment]
> It doesn't use the `requirements.txt`?

There are Aesara lines in there and those env files are used for the test pipelines..

**Comment by ricardoV94** [comment]
Changed the conda-envs

**Review by michaelosthege — APPROVED** [review]

**Pull request merged**